syntax = "proto3";
option cc_enable_arenas = true;

package Ydb.DynamicConfig.V1;

option java_package = "com.yandex.ydb.dynamic_config.v1";

import "ydb/public/api/protos/draft/ydb_dynamic_config.proto";

// Dynamic Config is the configuration applied to dynamic nodes without static configuration
// and is utilized by certain services, such as logging, which can update their settings
// seamlessly without requiring any restarts.

service DynamicConfigService {
    // Apply new version of config.
    // Overwrites previous version.
    // This call is idempotent:
    //  - Two subsequent identical calls will return success,
    //  - After applying next version another identical call will result in an error.
    // The field version within the YAML in the request must strictly be set to the current version increment by one.
    // The field cluster within the YAML should be identical to the one configured on the node used by the console tablet.
    rpc ApplyConfig(DynamicConfig.ApplyConfigRequest) returns (DynamicConfig.ApplyConfigResponse);

    // Get current configs.
    rpc GetConfig(DynamicConfig.GetConfigRequest) returns (DynamicConfig.GetConfigResponse);

    // Drop current config.
    // This call is idempotent.
    rpc DropConfig(DynamicConfig.DropConfigRequest) returns (DynamicConfig.DropConfigResponse);

    // Add volatile config.
    rpc AddVolatileConfig(DynamicConfig.AddVolatileConfigRequest) returns (DynamicConfig.AddVolatileConfigResponse);

    // Remove volatile config.
    rpc RemoveVolatileConfig(DynamicConfig.RemoveVolatileConfigRequest) returns (DynamicConfig.RemoveVolatileConfigResponse);

    // Resolve config for particular labels.
    rpc ResolveConfig(DynamicConfig.ResolveConfigRequest) returns (DynamicConfig.ResolveConfigResponse);

    // Resolve config for all possible labels combinations.
    rpc ResolveAllConfig(DynamicConfig.ResolveAllConfigRequest) returns (DynamicConfig.ResolveAllConfigResponse);
}
